build_modules/lib/src/meta_module_clean_builder.dart

```diff
@@ -0,0 +1,125 @@
+    for (var source in module.sources) {
+      assetToModule[source] = module;
+    }
+    for (var dep in module.directDependencies) {
+      var depMetaAsset = new AssetId(dep.package, 'lib/$metaModuleExtension');
+      if (!seenMetas.contains(depMetaAsset)) {
+        seenMetas.add(depMetaAsset);
+        if (await buildStep.canRead(depMetaAsset)) {
+          var depMeta = new MetaModule.fromJson(
+              json.decode(await buildStep.readAsString(depMetaAsset))
+                  as Map<String, dynamic>);
+          nextModules.addAll(depMeta.modules);
+        } else {
+          log.info('Unable to find meta module for dependency: $dep '
+              'Are you missing a dependency on package:${dep.package}?');
+        }
+      }
+    }
+  }
+  return dependentModules;
+}
+
+/// Merges the modules in a strongly connected component.
+///
+/// Note this will clean the module dependencies as the merge happens.
+/// The result will be that all dependencies are primary sources.
+Module _mergeComponent(
+    List<Module> connectedComponent, Map<AssetId, Module> assetToModule) {
+  var sources = new Set<AssetId>();
+  var deps = new Set<AssetId>();
+  // Sort the modules to deterministicly select the primary source.
+  var components = new SplayTreeSet<Module>(
+      (a, b) => a.primarySource.compareTo(b.primarySource))
+    ..addAll(connectedComponent);
+  var primarySource = components.first.primarySource;
+  for (var module in connectedComponent) {
+    sources.addAll(module.sources);
+    for (var dep in module.directDependencies) {
+      var depModule = assetToModule[dep];
+      if (depModule == null) continue;
+      // This dep is now merged into sources so skip it.
+      if (!components.contains(depModule)) deps.add(depModule.primarySource);
+    }
+  }
+  return new Module(primarySource, sources, deps);
+}
+
+/// Creates `.meta_module_clean` file for any Dart library.
+///
+/// This file contains information about the full computed
+/// module structure for the package with cleaned dependencies.
+class MetaModuleCleanBuilder implements Builder {
+  const MetaModuleCleanBuilder();
+
+  @override
+  final buildExtensions = const {
+    '$metaModuleExtension': const [metaModuleCleanExtension]
+  };
+
+  @override
+  Future build(BuildStep buildStep) async {
+    var assetToModule = await buildStep.fetchResource(_assetToModule);
+    var modules =
+        await _transitiveModules(buildStep, buildStep.inputId, assetToModule);
+    var connectedComponents = stronglyConnectedComponents<AssetId, Module>(
+        modules,
+        (m) => m.primarySource,
+        (m) => m.directDependencies
+            .map((d) => assetToModule[d])
+            .where((d) => d != null));
+    // Ensure deterministic output by sorting the modules.
+    var cleanModules = new SplayTreeSet<Module>(
+        (a, b) => a.primarySource.compareTo(b.primarySource));
+    for (var connectedComponent in connectedComponents) {
```

Comment:
could write this in a way that more directly expresses that we're interested the collection rather than side effects.

```dart
  Module merge(List<Module> c) => _mergeComponent(c, assetToModule);
  bool hasSourceInPackage(Module m) =>
      m.sources.any((s) => s.package == buildStep.inputId.package);
  new SplayTreeSet<Module>()
    ..addAll(connectedComponents.map(merge).where(hasSourceInPackage));
```